[pg] Remove columns, props and args from options

aravindet opened the pull request.

aravindet commented on src/pg/index.js:
```diff
@@ -35,65 +21,18 @@ export default (opts = {}) =>
-    }
-
-    setInterval(poll, pgOptions.pollInterval);
+    const myTransaction = new Transaction({ client, pgOptions });
```
First on naming: For me, this object is not a transaction, it is just a wrapper around the PG client. A transaction is what you do with SQL `BEGIN`, `COMMIT` and `ROLLBACK` statements. We will have to add support for transactions soon, so it is better to name this something else. I suggest naming it DbWrapper, so we do `const db = new DbWrapper(...)`, and you can rename the methods so it becomes `db.read()` rather than `db.dbRead()`.

Second, we should only accept a pool object (not client). This is required for both stability (clients disconnect eventually and don't reconnect automatically; we need this to last the lifetime of a server process) and performance (clients can only do one query at a time, no parallelism). I suggest we replace the `client` option here with `pgPool`.

Finally, just to make sure we are on the same page (no change required): this library will eventually need to get PG connections in 4 different ways. 1, 2 and 3 are based on the behavior of other PG libraries in NPM, such as migrations and graphile-worker; 4 is specific to us.

1. User does not provide anything. Library should create a Pool using PSQL environment variables. This is helpful when we need to write one-off scripts.
2. User provides the credentials at the time of server initialization (in the `.use(pg(...))` call), and the library uses these to create a Pool. Also helpful for one-off scripts.
3. User provides a [pool](https://node-postgres.com/api/pool) object at the time of server initialization (in the `.use(pg(...))` call). This is what we are supporting here.
4. We get a client in each query (in the `.read(query, options)` or `.write(change, options)` calls). We need this in Lego for proper transactions (atomicity) in Lego writes. We will implement it using a separate middleware, called `transaction` which might add an option like `graffyPgDbWrapper` to the options object. We can build this in another PR, this is already quite big.

annieDang replied:
yeah, the handling of transaction will be in the next PR

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+import sql, { raw, join } from 'sql-template-tag';
```
(minor) I try to avoid names like `helpers`, `utils` etc. This file seems to be about generating small clauses of SQL, I suggest calling it `clauses.js`?

aravindet commented on src/pg/sql/getMeta.js:
```diff
@@ -0,0 +1,15 @@
+    $ver: nowTimestamp,
+  });
+
+export const getMeta = (key, prefix, idCol) =>
```
This is great! Minor point, let's call it `getArgMeta`, to be consistent with all other read code where we have two branches, `id` and `arg`.

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+  return sql`to_jsonb("${raw(table)}")`;
+};
+
+export const colsAndValues = (data, versionCol) => ({
```
For consistency with other functions here which all start with `get` and return a PG SQL object, I suggest splitting this into two functions: `getInsertCols` and `getInsertVals`.

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+  values: join(Object.values(data).concat(nowTimestamp), ', '),
+});
+
+export const whereConditions = (data) =>
+  join(
+    Object.entries(data).map(([name, value]) => {
+      return sql`"${raw(name)}" = ${value}`;
+    }),
+  );
```
We can remove this unused function.

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+  );
+
+export const getUpdates = (data, options) => {
+  const { id, version } = options;
```
It's idCol and verCol, right?

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+
+export const nowTimestamp = sql`cast(extract(epoch from now()) as integer)`;
+
+export const getJsonBuildObject = (variadic) => {
```
:+1: Great idea.

aravindet commented on src/pg/sql/select.js:
```diff
@@ -1,47 +1,30 @@
-    ? 1
-    : Math.min(MAX_LIMIT, Math.max(0, limit || 0));
-
+  const clampedLimit = limit || MAX_LIMIT;
```
We need `Math.min(MAX_LIMIT, Math.max(0, limit || 0));` here, to clamp limits that are higher than MAX_LIMIT.

aravindet commented on src/pg/sql/helper.js:
```diff
@@ -0,0 +1,47 @@
+  return sql`to_jsonb("${raw(table)}")`;
+};
+
+export const colsAndValues = (data, versionCol) => ({
+  cols: join(
+    Object.keys(data)
+      .map((col) => raw(`"${col}"`))
+      .concat(raw(`"${versionCol}"`)),
+    ', ',
+  ),
+  values: join(Object.values(data).concat(nowTimestamp), ', '),
+});
+
+export const whereConditions = (data) =>
+  join(
+    Object.entries(data).map(([name, value]) => {
+      return sql`"${raw(name)}" = ${value}`;
+    }),
+  );
+
+export const getUpdates = (data, options) => {
+  const { id, version } = options;
+  return join(
+    Object.entries(data)
+      .filter(([name]) => name !== id)
+      .map(([name, value]) => {
+        return sql`"${raw(name)}" = ${value}`;
+      })
+      .concat(sql`"${raw(version)}" =  ${nowTimestamp}`),
+    ', ',
+  );
+};
```
I think we can delete these 3 unused functions, there is more complete implemetations in getArgSql.js and upsert.js

annieDang commented on src/pg/sql/clauses.js:
```diff
@@ -0,0 +1,48 @@
+
+export const getSelectCols = (table) => {
+  // TODO: When we have a query object, get only the requested columns.
+  return sql`to_jsonb("${raw(table)}")`;
```
FYI, It has to be `to_jsonb` to make the whole concatenation works.

aravindet commented on src/pg/db.js:
```diff
@@ -19,24 +19,25 @@ import { put, patch } from './sql/index.js';
   // client is pgClient
-  constructor({ client, pgOptions }) {
-    this.client = client;
+  constructor({ pgOptions }) {
```
Shouldn't it be `constructor(pgOptions)`? I don't expect any other options to be passed to pg module.

aravindet commented on src/pg/db.js:
```diff
@@ -106,18 +91,10 @@ export class DbWrapper {
   }
 
   async write(change, opts = this.pgOptions) {
-    async function writeSql(query, client) {
```
:+1: Good idea to move these into the PgDb class.

aravindet commented on src/pg/db/config.js:
```diff
@@ -0,0 +1,27 @@
+import debug from 'debug';
```
(minor) To use the debug library, you must do this:

```js
import debug from 'debug';

const log = debug('graffy:pg:config'); // add a tag for all the lines logged from this file.

log(...); // This is the function you call to log things.
```

Later, when running your program, you will see no log lines. To enable them, you have to set an environment variable:

```sh
DEBUG=graffy:pg:*
```

You can also do `DEBUG=*` but it will be very noisy (Graffy is not the only library using `debug`).

annieDang replied:
what is the convention here?
at this particular file I use 
`debug(`graffy:pg:error`)(
          `Error in Postgres ConnectionPool: ${err.message}`,
        );` 
isn't it equivalent with 
`const log = debug('graffy:pg:error');
log(`Error in Postgres ConnectionPool: ${err.message}`)`
?

aravindet replied:
The convention used elsewhere in graffy is to define a log function.

aravindet commented on src/pg/db/pool.js:
```diff
@@ -0,0 +1,81 @@
+
+  async write(query) {
+    return this.query(query, 'write');
+  }
```
(minor) We can avoid the "type parameter" with a small refactor:

```js
  async query(query) {
    const start = Date.now();
    query.rowMode = 'array';
    const res = await this.pool.query(query);
    const duration = Date.now() - start;
    log(type, `${{ query: query.text, duration, rows: res.rowCount }}`);

    return res;
  }

  async read(query) {
      const result = await this.query(query);
      result = result.rows.flat();
      // Each row is an array, as there is only one column returned.
      log('Read result', result);
      return result;
  }

  async write(query) {
      const result = this.query(query);
      log('Rows written', res.rowCount);
      return res.rowCount;
  }
```

annieDang replied:
this is better, the reason why i used type is the log of original version:
`log(type, `${{ query: query.text, duration, rows: res.rowCount }}`);`
But it is not necessary so it will look like your suggestion!

aravindet commented on src/pg/test/db/dbWrite.test.js:
```diff
@@ -0,0 +1,86 @@
+
+  beforeEach(async () => {
+    jest.useFakeTimers();
+    PgDb.prototype = { write: mockQuery, getClient: mockClient };
```
Shouldn't the getClient mock be a function that returns the mockClient object?

But in any case, as you have mocked the .write, it looks like `getClient` never gets called. This may be unnecessary.

annieDang commented:
It works properly with lego. `sample.ts` needs new opts to make it work tho.
`const options = {
        opts: {
            idCol: 'id',
            verCol: '_version',
        },
    };`

aravindet commented on src/pg/db.js:
```diff
@@ -0,0 +1,124 @@
+    return slice(results, rootQuery).known || [];
+  }
+
+  async write(change, opts = this.pgOptions) {
```
This is not a good approach. Operation (read/write) options and module-level options (pgOptions) are completely different things.

- Module-level options are things like connection parameters, column names for ID and Version columns, etc. They remain unchanged from the time the server starts to the time it is shut down.
- Operation-level options are things like authentication tokens, transaction ID, transaction client, flags about whether to read from cache, etc. These change from one request to the next.

The `options = this.pgOptions` makes no sense. I would suggest not to define opts in this PR at all, just use this.pgOptions directly everywhere.

aravindet replied:
Hold on, I might have misunderstood this.

aravindet commented on src/pg/db/pool.js:
```diff
@@ -0,0 +1,76 @@
+import { config as defaultConfig } from './config';
+
+import debug from 'debug';
+const log = (...text) => debug(`graffy:pg:${text.shift()}`)(text);
```
```suggestion
const log = debug(`graffy:pg:pool`);
```

"error" is not a good tag for the debug library. The convention is to use tags to isolate the part of the application. The `debug` library that we use does not have a concept of log levels (error, warn, info).

aravindet commented on src/pg/test/sql/select.test.js:
```diff
@@ -0,0 +1,86 @@
+import { selectByArgs, selectByIds } from '../../sql/select.js';
```
The tests are great! Thank you!

One minor point on the naming, I prefer to use names that are easily grep-able, so I can specify them on the command line.

Here I might suggest:
- describe('pg_select')
- tests:
  - byArgs_first
  - byIds
  - byArgs_order_first
  - byArgs_order_before

aravindet commented on src/pg/test/sql/clauses.test.js:
```diff
@@ -0,0 +1,43 @@
+import sql, { raw } from 'sql-template-tag';
```
minor, naming:

  -  describe('sql_helpers')
  -  tests:
        - insert_cols_and_values
        - updates
        - json_build_object
        - select_columns

aravindet commented:
I have updated it to fix the issues. Primarily, I did not see any point in two classes (PgDb and DbWrapper), combined them into one (Db). I have also removed the default configs (we just use the pg library defaults) and some code that was catching and printing errors in config.js and PgDb.js (seems to be aimed towards debugging).

Also removed simple-pg with this PR.

Note that you have to update the Lego code:

Firstly,
```diff
-import pg from '@graffy/pg';
+import { pg } from '@graffy/pg';
```

Secondly,
```diff
- store.use(name, pg({ opts }));
+ store.use(name, pg(opts));
```

Third, the options should be `idCol` and `verCol`, not `id` and `version`.

Fourth, the `connection` option can be a plain JS object like `{ host, user, password, port, database }` (as it is now), but it can also be:
- a pg.Pool object
- a pg.Client object
- a string, like "postgresql://dbuser:secretpassword@database.server.com:3211/mydb"

annieDang commented:
Thank you for the cleaning up! Let's merge this so we can move on to the next PR! :)
